docs/proposals/lifecycle_events.md

```diff
@@ -0,0 +1,77 @@
+# Support execution of the commands based on lifecycle events.
+
+## Abstract
+Add support for the lifecycle events which can be defined in 2.0.0 devfiles.
+
+Our proposed solution is to perform the preStart and postStart events as a part of the **odo push** command, and the preStop and Poststop events as a part of the **odo delete** command.
+
+Lifecycle bindings for specific events - https://github.com/devfile/kubernetes-api/issues/32
+
+
+## Motivation
+Devfile support commands that can be triggered based on dev lifecycle events. Odo will need to support/execute these commands at appropriate times within the flow.
+
+With the implementation of livecycle events, stack creators will be able to leverage all the capabilities of devfiles when building the correct experience/features for their stacks.
+
+## User Stories
+Each lifecycle event would be suited to individual user stories. We propose that at least one issue is made for each of them for tracking and implementation purposes.
+
+An issue has already been created for postStart: [container initiliazation support](https://github.com/openshift/odo/issues/2936)
+
+## Design overview
+
+The lifecycle events that have been implemented in Devfile 2.0.0 are generic. So their effect/meaning might be slightly different when applying them to a workspace (Che) or a single project (odo). However, this proposal will aim to provide a similar user experience within the scope of a single project.
+
+Our proposal is that the preStart and postStart events are designed to run as a part of **odo push**, and that the preStop and postStop events run as a part of **odo delete**.
+
+
+
+### The flow for **odo push** including the *preStart* and *postStart* lifecycle events will be as follows:
+ - preStart - Devfile command gets translated to entrypoint for specified init container - one time only
+ - Container initialisation (as done today)
+ - postStart - Exec the specified command(s) in the container - one time only
+ - Rest of the command execution (as we do today - exec commands for build/run/test/debug)
+
+**preStart**:
+ - The preStart command(s) that are specified in the Devfile will be translated into entry points for their specified containers, and added to the pod spec as init containers. 
+ - These will only run on the first odo push, or when doing a force push (odo push -f). 
+ - These are commands that you’d want to run before the main containers are created. 
+
+    **Note: We need some good use cases to understand when the user would run commands as preStart opposed to postStart.**
+
+**Component Initialization**
+ - After the preStart event has completed, the containers specified in the Devfile are initialised and created if the component doesn’t already exist.
+
+**postStart**
```

Comment:
1. Consider to add some implementation details on this design, e.g. before starting a given container in odo, we scan for all the events `postStart` events defined in the devfile and find out all the commands that uses the same container. This is the logic to identify the list of `postStart` event that is needed to run for a given container.
2. Add error handling. What if the `postStart` command failed to execute, how do we handle it and how do we surface that error to the user.
3. The type of container (i.e. the command group that links to a given container) may affect the behviour, e.g. the `postStart` only gets executed in the first `odo push` in most cases for build containers (containers used by the build command) since build containers will stay up and running after the first push. However, for `run` containers, the container may or may not get restarted depending on the `restart` flag. So the `postStart` may gets executed again for `run` containers. What about containers used by `debug` and `test` group of commands? We may want to clarify the behaviour in this proposal.

Reply:
I think point#1 and point#3 have not been addressed yet.

Reply:
I understand the premise of your point, but I have been struggling to think of any specific use cases where this would be a driving factor.
We (@EnriqueL8, @neeraj-laad, and I) have discussed the potential to be more granular in the solution by identifying which container specifically we are initialising, and whether or not there is a postStart command associated with it. This way we could better utilise the postStart commands and use it in cases where only come of the containers are being initialised.


A potential flow could be to check for postStart commands associated with a specific container when inititialising it, storing those commands in an Object(?) and then executing only those postStart commands when the containers have all finished initialising, in the order that they are defined in within the devfile. However, this isn't necessary for the basic implementation of the postStart event. There could be room for such a feature though, and a case could be presented to argue the validity of changing the flow of `odo push` in order to accommodate such cases in a future implementation.

I will update the proposal to reflect this though, as it was definitely something we hadn't considered.